File: server/set_schedule.py
#!/usr/bin/env python3
"""
AlertMed - Set Schedule & Rotate Pin 5 Servo
Usage: python set_schedule.py <time_HH:MM> [pin5_degree]
       python set_schedule.py <time_HH:MM> --degree <pin5_degree>
       python set_schedule.py --degree <pin5_degree>
"""

import sys
import re
import json
import urllib.request
import urllib.error

TILE_PIN = 12
MEDICINE_NAME = "Paracetamol"
SCHEDULE_TYPE = "once_daily"
DAYS = "daily"

SERVO_ESP32_URL = "http://10.180.37.20:8080"
SERVER_URL = "http://127.0.0.1:3000"

def rotate_pin5_servo(degree):
    try:
        deg = int(degree)
    except ValueError:
        print(f"[ERROR] Invalid degree '{degree}'. Must be a number between 0 and 180.")
        return False

    if deg < 0 or deg > 180:
        print(f"[ERROR] Invalid degree {deg}. Must be between 0 and 180.")
        return False

    print(f"\n[SERVO PIN 5] Sending command to rotate Pin 5 servo to {deg} degrees...")
    url = f"{SERVO_ESP32_URL}/spin?pin=5&degree={deg}&duration=1000"
    try:
        req = urllib.request.Request(url, method='GET')
        with urllib.request.urlopen(req, timeout=10) as resp:
            data = json.loads(resp.read().decode())
            print(f"[SUCCESS] Pin 5 servo rotated to {deg} degrees! ({data.get('message', 'OK')})")
            return True
    except Exception as e:
        print(f"[ERROR] Failed to rotate Pin 5 servo at {SERVO_ESP32_URL}: {e}")
        return False

def main():
    args = sys.argv[1:]
    if len(args) < 1:
        print("=" * 60)
        print("          AlertMed - Set Schedule & Servo")
        print("=" * 60)
        print("Usage: python set_schedule.py <time_HH:MM> [pin5_degree]")
        print("       python set_schedule.py <time_HH:MM> --degree <pin5_degree>")
        print("       python set_schedule.py --degree <pin5_degree>")
        print("")
        print("Examples:")
        print("  python set_schedule.py 22:50")
        print("  python set_schedule.py 22:50 90")
        print("  python set_schedule.py 22:50 --degree 180")
        print("  python set_schedule.py --degree 45")
        print("=" * 60)
        sys.exit(1)

    time_val = None
    pin5_degree = None

    i = 0
    while i < len(args):
        arg = args[i]
        if arg in ('--degree', '-d', '--deg'):
            if i + 1 < len(args):
                pin5_degree = args[i + 1]
                i += 1
        elif not time_val and re.match(r'^([01]\d|2[0-3]):([0-5]\d)$', arg):
            time_val = arg
        elif pin5_degree is None and arg.isdigit() and ':' not in arg:
            pin5_degree = arg
        i += 1

    # 1. Rotate Pin 5 servo if degree is provided
    if pin5_degree is not None:
        rotate_pin5_servo(pin5_degree)

    # 2. Add schedule if time is provided
    if time_val:
        print(f"\n[SCHEDULE] Setting up Tile (Pin {TILE_PIN}) with {MEDICINE_NAME}...")
        try:
            req = urllib.request.Request(
                f"{SERVER_URL}/api/update-medicine",
                data=json.dumps({'pin': TILE_PIN, 'name': MEDICINE_NAME}).encode('utf-8'),
                headers={'Content-Type': 'application/json'},
                method='POST'
            )
            with urllib.request.urlopen(req, timeout=3) as resp:
                data = json.loads(resp.read().decode())
                print(f"[SUCCESS] {data.get('message', 'Medicine updated')}")
        except Exception as e:
            print(f"[ERROR] Could not update medicine: {e}")

        print(f"[SCHEDULE] Adding schedule at {time_val}...")
        try:
            req = urllib.request.Request(
                f"{SERVER_URL}/api/schedule",
                data=json.dumps({
                    'pin': TILE_PIN,
                    'time': time_val,
                    'schedule_type': SCHEDULE_TYPE,
                    'days': DAYS
                }).encode('utf-8'),
                headers={'Content-Type': 'application/json'},
                method='POST'
            )
            with urllib.request.urlopen(req, timeout=3) as resp:
                data = json.loads(resp.read().decode())
                print(f"[SUCCESS] Schedule {time_val} added to tile {TILE_PIN}")
        except Exception as e:
            print(f"[ERROR] Could not add schedule: {e}")

if __name__ == "__main__":
    main()

File: server/set_schedule.js
const axios = require('axios');

// ==========================================
// ⚙️ CONFIGURE YOUR MEDICINE & HARDWARE DETAILS
// ==========================================
const TILE_PIN = 12;                     // Default pin number for the scheduled medicine tile
const MEDICINE_NAME = "Paracetamol";     // Name of the medicine
const SCHEDULE_TYPE = "once_daily";      // Type of schedule
const DAYS = "daily";                    // Days to dispense (e.g., "daily" or "Mon, Wed, Fri")

const SERVO_ESP32_URL = 'http://10.180.37.20:8080';  // Servo ESP32 base URL
const SERVER_URL = 'http://127.0.0.1:3000';          // Node.js server base URL
// ==========================================

async function rotatePin5Servo(degree) {
    const deg = parseInt(degree, 10);
    if (isNaN(deg) || deg < 0 || deg > 180) {
        console.error(`[ERROR] Invalid degree '${degree}'. Degree must be between 0 and 180.`);
        return false;
    }

    console.log(`\n[SERVO PIN 5] Sending command to rotate Pin 5 servo to ${deg} degrees...`);
    try {
        const res = await axios.get(`${SERVO_ESP32_URL}/spin?pin=5&degree=${deg}&duration=1000`, { timeout: 10000 });
        if (res.data && res.data.success !== false) {
            console.log(`[SUCCESS] Pin 5 servo rotated to ${deg} degrees!`);
            return true;
        } else {
            console.warn(`[WARNING] ESP32 response:`, res.data);
            return false;
        }
    } catch (err) {
        console.error(`[ERROR] Failed to rotate Pin 5 servo at ${SERVO_ESP32_URL}:`, err.message);
        return false;
    }
}

async function setMedicineAndSchedule() {
    const rawArgs = process.argv.slice(2);

    if (rawArgs.length < 1) {
        console.log("=========================================================");
        console.log("             AlertMed - Set Schedule & Servo             ");
        console.log("=========================================================");
        console.log("Usage: node set_schedule.js <time_HH:MM> [pin5_degree]");
        console.log("       node set_schedule.js <time_HH:MM> --degree <pin5_degree>");
        console.log("       node set_schedule.js --degree <pin5_degree>");
        console.log("");
        console.log("Examples:");
        console.log("  node set_schedule.js 22:50");
        console.log("  node set_schedule.js 22:50 90");
        console.log("  node set_schedule.js 22:50 --degree 180");
        console.log("  node set_schedule.js --degree 45");
        console.log("=========================================================");
        process.exit(1);
    }

    let time = null;
    let pin5Degree = null;

    // Parse arguments supporting both positional and flags
    for (let i = 0; i < rawArgs.length; i++) {
        const arg = rawArgs[i];
        if (arg === '--degree' || arg === '-d' || arg === '--deg') {
            if (i + 1 < rawArgs.length) {
                pin5Degree = rawArgs[i + 1];
                i++;
            }
        } else if (!time && /^([01]\d|2[0-3]):([0-5]\d)$/.test(arg)) {
            time = arg;
        } else if (!pin5Degree && !isNaN(parseInt(arg, 10)) && !arg.includes(':')) {
            pin5Degree = arg;
        }
    }

    // 1. If degree parameter is specified, rotate Pin 5 servo
    if (pin5Degree !== null) {
        await rotatePin5Servo(pin5Degree);
    }

    // 2. If time is specified, register medicine and set schedule on server
    if (time) {
        // Basic time format validation
        const timeRegex = /^([01]\d|2[0-3]):([0-5]\d)$/;
        if (!timeRegex.test(time)) {
            console.error("[ERROR] Invalid time format. Please use HH:MM (e.g., 08:30 or 14:45)");
            process.exit(1);
        }

        try {
            console.log(`\n[SCHEDULE] Setting up Tile (Pin ${TILE_PIN}) with ${MEDICINE_NAME}...`);
            
            // 1. Update medicine details on tile
            const updateRes = await axios.post(`${SERVER_URL}/api/update-medicine`, {
                pin: TILE_PIN,
                name: MEDICINE_NAME
            });
            
            if (updateRes.data.success) {
                console.log(`[SUCCESS] ${updateRes.data.message}`);
            } else {
                console.log(`[FAILED] ${updateRes.data.message}`);
            }

            // 2. Set the schedule for the tile
            console.log(`[SCHEDULE] Adding schedule at ${time}...`);
            const scheduleRes = await axios.post(`${SERVER_URL}/api/schedule`, {
                pin: TILE_PIN,
                time: time,
                schedule_type: SCHEDULE_TYPE,
                days: DAYS
            });
            
            if (scheduleRes.data.success) {
                console.log(`[SUCCESS] ${scheduleRes.data.message}`);
            } else {
                console.log(`[FAILED] ${scheduleRes.data.message}`);
            }

        } catch (error) {
            if (error.response) {
                console.error(`[ERROR] Server rejected request:`, error.response.data);
            } else {
                console.error(`[ERROR] Failed to connect to server at ${SERVER_URL}:`, error.message);
            }
        }
    }
}

setMedicineAndSchedule();
